schemachanger: refactor subzone config elements

annrpom opened the pull request:
### config/zonepb: add subzoneSpan-related methods
This patch adds subzoneSpan-related methods that aid in configuring
subzone zcs in the declarative schemachanger.

Epic: none

Release note: None

---

### schemachanger: extend Catalog interface to support subzone writes
This patch extends the `Catalog` interface to support getting and
writing a zone config. This is necessary for us to be able to get
the parent zone config, apply all subzone changes, and then write
to kv in `immediateState`.

Epic: none

Release note: None

---

### schemachanger: refactor subzone config elements
This patch refactors how subzoneSpans are handled in the declarative schema changer. We now filter out subzoneSpans that do not relate to the index/partition zone config element it is nested within. This makes discarding these subzone elements feasible.

In order to support zone config changes where the target's subzone spans -- along with other targets' subzone spans -- are modified, we make those changes a side-effect. That is:

```
...
ALTER PARTITION default OF TABLE db.person CONFIGURE ZONE USING gc.ttlseconds = 1;
-> PartitionZoneConfig{default}

ALTER PARTITION australia OF TABLE db.person CONFIGURE ZONE USING gc.ttlseconds = 2;
-> PartitionZoneConfig{default}
and
-> PartitionZoneConfig{australia}
```

Informs: #133158 
Epic: none

Release note: None

---

### schemachanger: refactor add/drop zc elements
This patch separates marking zc elements as dropped from
the add direction in the builder as subzone configs will have
a mixed state of drop subzone + add side effects.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133879)
<!-- Reviewable:end -->

annrpom commented:
oh oops the spans will have to be sorted better than that

annrpom commented:
woof ok -- so this is almost there. the problem now is from a situation like https://github.com/cockroachdb/cockroach/blob/f2f059d083bc3acea052134b3bc79b86bd5ffeba/pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/testdata/partitions#L2-L39


where just updating australia's subzone spans is not going to work; we need to add support to configure pre-existing subzone configs' subzone spans as a side effect

so:

`ALTER PARTITION default OF TABLE db.person CONFIGURE ZONE USING gc.ttlseconds = 1;` will have
- `PartitionZoneConfig{default}` 

`ALTER PARTITION australia OF TABLE db.person CONFIGURE ZONE USING gc.ttlseconds = 2;` will have 
- `PartitionZoneConfig{australia}`
and
- `PartitionZoneConfig{default}` (just to update its subzone spans accordingly)

or something of the sort

annrpom commented on pkg/config/zonepb/zone.go:
```diff
@@ -1153,6 +1154,30 @@ func (z *ZoneConfig) DeleteSubzone(indexID uint32, partition string) bool {
+		}
+	}
+
+	z.SubzoneSpans = append(filteredSpans, subzoneSpans...)
```
TODO: leave a comment about nearly sorted thing

annrpom commented:
arg i tried to do the optimization https://gist.github.com/annrpom/8a5012562c8e0682b8f995e6e3eb46c8

and it's ok, but not in transactions for some reason -- i left a TODO and put ^ gist up as a future task

fqazi reviewed (commented):
Great work @annrpom, this is coming along nicely.

> Reviewed 6 of 14 files at r1, 23 of 23 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-:-OArfEmPEllf6JjA-vDj:bfun59a)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1220 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OArdU174Z_RCBeGRSoa:-OArdU174Z_RCBeGRSob:b-jxcg0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e9a20430e96937abf4a71667c8df5bee3c004b6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1220)):*
> ```Go
> // for a lookup of which subzone spans a particular subzone is referred to by.
> func getAffectedSubzoneSpansWithIdx(
> 	subzonesSlice int, newSubzoneSpans []zonepb.SubzoneSpan,
> ```

Nit: maybe call subzoneSlices something like numSubZones  (or something more descriptive)
___
*[`pkg/sql/schemachanger/scexec/exec_immediate_mutation.go` line 239 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OArdqPo6CH_HvT_w-wt:-OArdqPo6CH_HvT_w-wu:b-x7biyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e9a20430e96937abf4a71667c8df5bee3c004b6/pkg/sql/schemachanger/scexec/exec_immediate_mutation.go#L239)):*
> ```Go
> 	for _, zcToUpdate := range s.modifiedZoneConfigs {
> 		if zcToUpdate.isSubzoneConfig {
> 			// TODO(before merge): unsatisfied with this; would it be better to
> ```

Yeah it might be cleaner to eliminate the overload


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133879)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1220 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OArdU174Z_RCBeGRSoa:-OArwxSa6gjJZ2SUFWFB:b-86jyed) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e9a20430e96937abf4a71667c8df5bee3c004b6/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1220)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: maybe call subzoneSlices something like numSubZones  (or something more descriptive)
</blockquote></details>

done -- the naming of the method was also a remnant of something i tried doing that didn't pan out; so, i changed that as well
___
*[`pkg/sql/schemachanger/scexec/exec_immediate_mutation.go` line 239 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OArdqPo6CH_HvT_w-wt:-OArxBSp-WZ5gaewFNal:b-fnrx58) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e9a20430e96937abf4a71667c8df5bee3c004b6/pkg/sql/schemachanger/scexec/exec_immediate_mutation.go#L239)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Yeah it might be cleaner to eliminate the overload
</blockquote></details>

done oki


<!-- Sent from Reviewable.io -->

annrpom committed:
schemachanger: refactor subzone config elements

This patch refactors how subzoneSpans are handled in the declarative schema
changer. We now filter out subzoneSpans that do not relate to the
index/partition zone config element it is nested within. This makes discarding
these subzone elements feasible.

In order to support zone config changes where the target's subzone spans --
along with other targets' subzone spans -- are modified, we make those changes
a side-effect. That is:

```
...
ALTER PARTITION default OF TABLE db.person CONFIGURE ZONE USING gc.ttlseconds = 1;
-> PartitionZoneConfig{default}

ALTER PARTITION australia OF TABLE db.person CONFIGURE ZONE USING gc.ttlseconds = 2;
-> PartitionZoneConfig{default}
and
-> PartitionZoneConfig{australia}
```

Informs: #133158
Epic: none

Release note: None

annrpom committed:
config/zonepb: add subzoneSpan-related methods

This patch adds subzoneSpan-related methods that aid in configuring
subzone zcs in the declarative schemachanger.

Epic: none

Release note: None

annrpom committed:
schemachanger: refactor add/drop zc elements

This patch separates marking zc elements as dropped from
the add direction in the builder as subzone configs will have
a mixed state of drop subzone + add side effects.

Epic: none

Release note: None

fqazi reviewed (approved):
This look great @annrpom. Some very minor comments, but all of this much neater.

> Reviewed 29 of 29 files at r4, 29 of 29 files at r5, 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-:-OBS78dK4uwXecxFvOIr:bqf7s36)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OBS6mJg5hfwwNknpnaB:-OBS6mJg5hfwwNknpnaC:b-1sjz38) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2cd44dd98141f4e2b7cf6d49f6134224356490e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L91)):*
> ```Go
> 			return
> 		}
> 		toDrop, sideEffects := zco.getZoneConfigElemForDrop(b)
> ```

Nit: Use a more informative name here for sideEffects
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/partition_zone_config.go` line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OBS5cC54M-C0IOMvyME:-OBS5cC60cN1CsvOdJMn:b-1jojro) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b336f8220757889bb6f83ec7907ff85a1564cc4/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/partition_zone_config.go#L104)):*
> ```Go
> 		zc, err = pzo.getInheritedDefaultZoneConfig(b)
> 	} else {
> 		//zc, err = pzo.tableZoneConfigObj.getZoneConfig(b, false /* inheritDefaultRange */)
> ```

Nit: Clean up
___
*[`pkg/config/zonepb/zone.go` line 1186 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OBS5NDQ5CwDcfyKKGQv:-OBS5NDQ5CwDcfyKKGQw:b-do8wac) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b336f8220757889bb6f83ec7907ff85a1564cc4/pkg/config/zonepb/zone.go#L1186)):*
> ```Go
> // DeleteSubzoneSpans deletes all given SubzoneSpan from z.
> func (z *ZoneConfig) DeleteSubzoneSpans(spansToDelete []SubzoneSpan) {
> 	toDeleteMap := make(map[string]bool, len(spansToDelete))
> ```

Nit: You could just make map[string]struct{}


<!-- Sent from Reviewable.io -->

annrpom committed:
schemachanger: extend Catalog interface to support subzone writes

This patch extends the `Catalog` interface to support getting and
writing a zone config. This is necessary for us to be able to get
the parent zone config, apply all subzone changes, and then write
to kv in `immediateState`.

Epic: none

Release note: None

annrpom reviewed (commented):
as discussed (offline), I added 1 more change: we now gather all subzone config changes on a table zone config before writing it to kv.  this prevents the span config watcher from observing partial states of subzone configurations (in particular: partial states of subzone span changes), which could lead to a panic in the span config translator  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-:-OBT1-2pENs4ml0pRaeP:bw6l50x)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/config/zonepb/zone.go` line 1186 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OBS5NDQ5CwDcfyKKGQv:-OBSN6HG3_Sj-noiLCCI:b-wmjmqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b336f8220757889bb6f83ec7907ff85a1564cc4/pkg/config/zonepb/zone.go#L1186)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: You could just make map[string]struct{}
</blockquote></details>

done oop yeah save mem; ty
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OBS6mJg5hfwwNknpnaB:-OBSR9Zk-iVO7eXWJ4l_:b-hawy01) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2cd44dd98141f4e2b7cf6d49f6134224356490e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/configure_zone.go#L91)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Use a more informative name here for sideEffects
</blockquote></details>

done  -- added a blurb as well
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/partition_zone_config.go` line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/133879#-OBS5cC54M-C0IOMvyME:-OBT-m176qTZIl3G-SOZ:b-8w0otu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b336f8220757889bb6f83ec7907ff85a1564cc4/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/partition_zone_config.go#L104)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Clean up
</blockquote></details>

done oop ty


<!-- Sent from Reviewable.io -->

annrpom commented:
TFTR! ('-')7 

bors r+
